src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1123 @@
+// limitations under the License.
+
+#include "decoder/lattice-faster-decoder-combine.h"
+#include "lat/lattice-functions.h"
+
+namespace kaldi {
+
+template<typename Token>
+BucketQueue<Token>::BucketQueue(BaseFloat best_cost_estimate,
+                                BaseFloat cost_scale) :
+    cost_scale_(cost_scale) {
+  // NOTE: we reserve plenty of elements to avoid expensive reallocations
+  // later on. Normally, the size is a little bigger than (adaptive_beam +
+  // 5) * cost_scale.
+  int32 bucket_size = 100;
+  buckets_.resize(bucket_size);
+  bucket_storage_begin_ = std::floor((best_cost_estimate - 15) * cost_scale_);
+  first_occupied_vec_index_ = bucket_size;
+}
+
+template<typename Token>
+void BucketQueue<Token>::Push(Token *tok) {
+  int32 bucket_index = std::floor(tok->tot_cost * cost_scale_);
+  size_t vec_index = static_cast<size_t>(bucket_index - bucket_storage_begin_);
+  if (vec_index >= buckets_.size()) {
+    int32 margin = 10;  // a margin which is used to reduce re-allocate
+                        // space frequently
+    // A cast from unsigned to signed type does not generate a machine-code
+    // instruction
+    if (static_cast<int32>(vec_index) > 0) {
+      KALDI_WARN << "Have to reallocate the BucketQueue. Maybe need to reserve"
+                 << " more elements in constructor. Push back.";
+      buckets_.resize(static_cast<int32>(vec_index) + margin);
+    } else {  // less than 0
+       KALDI_WARN << "Have to reallocate the BucketQueue. Maybe need to reserve"
+                 << " more elements in constructor. Push front.";
+       int32 increase_size = - static_cast<int32>(vec_index) + margin;
+       buckets_.resize(buckets_.size() + increase_size);
+       // translation
+       for (size_t i = buckets_.size() - 1; i >= increase_size; i--) {
+         buckets_[i].swap(buckets_[i - increase_size]);
+       }
+       bucket_storage_begin_ = bucket_storage_begin_ - increase_size;
+       vec_index = static_cast<int32>(vec_index) + increase_size;
+       first_occupied_vec_index_ = vec_index;
+    }
+  }
+  tok->in_queue = true;
+  buckets_[vec_index].push_back(tok);
+  if (vec_index < first_occupied_vec_index_)
+    first_occupied_vec_index_ = vec_index;
+}
+
+template<typename Token>
+Token* BucketQueue<Token>::Pop() {
+  int32 vec_index = first_occupied_vec_index_;
+  while (vec_index < buckets_.size()) {
+    Token* tok = buckets_[vec_index].back();
+    // Remove the best token
+    buckets_[vec_index].pop_back();
+
+    if (buckets_[vec_index].empty()) { // This bucket is empty. Update vec_index
+      int32 next_vec_index = vec_index + 1;
+      for (; next_vec_index < buckets_.size(); next_vec_index++) {
+        if (!buckets_[next_vec_index].empty()) break;
+      }
+      vec_index = next_vec_index;
+      first_occupied_vec_index_ = vec_index;
+    }
+
+    if (tok->in_queue) {  // This is a effective token
+      tok->in_queue = false;
+      return tok;
+    }
+  }
+  return NULL;
+}
```

Comment:
I just realized something about Pop(), which is a way to simplify it and possibly make it slightly more efficient.
the idea is to change the semantics of first_occupied_vec_index_, and rename it.  (obviously, document it).  E.g.
```
  /// first_nonempty_bucket_index_ is an integer in the range [0, buckets_.size() - 1]
  /// which is not larger than the index of the first nonempty element of buckets_.
  int32 first_nonempty_bucket_index_;
```
For clarity, I would like to get rid of the distinction between vec_index and bucket_index, which
makes things unclear.  I'd like to have a formula where:
```
  size_t bucket_index = std::floor(tok->tot_cost * cost_scale_) + bucket_offset_;
```
so you'd rename bucket_storage_begin_ to bucket_offset_ and negate it.


Then Pop() would be simplified to:
 ``` 
 Token *Pop() {
   if (!buckets_[first_nonempty_bucket_index_].empty()) {
       Token *ans = buckets_[first_nonempty_bucket_index_].back();
        buckets_[first_nonempty_bucket_index_].pop_back();
        return ans;
    }
    // and the rest of the code goes here.
 }
```

Reply:
Sorry, dan, I didn't get the difference between "first_nonempty_bucket_index_" and "first_occupied_vec_index_"

Dan, with the new design, I think the Pop() function still looks like current code. 
We cannot return the token directly as you wrote
```
if (!buckets_[first_nonempty_bucket_index_].empty()) {
  Token *ans = buckets_[first_nonempty_bucket_index_].back();
  buckets_[first_nonempty_bucket_index_].pop_back();
  return ans;
}
```
Because we cannot guarantee the 'ans' is an effective token.  We have to peek the next token.
```
while (first_nonempty_bucket_index_ < buckets_.size()) {
  Token *tok = buckets_[first_nonempty_bucket_index_].back();
  if (tok->in_queue) break;
  buckets_[first_nonempty_bucket_index_].pop_back();  //remove the invalid token
  if (buckets_[first_nonempty_bucket_index_].empty()) {
    first_nonempty_bucket_index_++;
    for (; first_nonempty_bucket_index_ < buckets_.size(); first_nonempty_bucket_index_++) {
      if(!bucket_[first_nonempty_bucket_index_].empty()) break;
    }
  }
}
```
At the same time, we still need "first_nonempty_bucket_index == buckets_.size()" to determinize the BucketQueue is empty at the beginning of the function.
```
if (first_nonempty_bucket_index == buckets_.size()) {  // no token left
  return NULL:
]
```